File: benchmarks/fixtures/webapp_ts/src/models/session.ts
/** Session model. */
import { getLogger } from '../utils/helpers';

const logger = getLogger("models.session");

export interface SessionRecord {
    id: string;
    userId: string;
    tokenHash: string;
    ipAddress: string;
    createdAt: number;
    expiredAt: number | null;
}

export class Session {
    public readonly id: string;
    public readonly userId: string;
    public readonly tokenHash: string;
    public readonly ipAddress: string;
    public readonly createdAt: number;
    public expiredAt: number | null;

    constructor(data: SessionRecord) {
        this.id = data.id;
        this.userId = data.userId;
        this.tokenHash = data.tokenHash;
        this.ipAddress = data.ipAddress;
        this.createdAt = data.createdAt;
        this.expiredAt = data.expiredAt;
    }

    isActive(): boolean {
        return this.expiredAt === null;
    }

    expire(): void {
        this.expiredAt = Date.now();
        logger.info(`Session ${this.id} expired`);
    }

    static create(userId: string, tokenHash: string, ip: string): Session {
        return new Session({
            id: `sess-${Date.now()}`,
            userId,
            tokenHash,
            ipAddress: ip,
            createdAt: Date.now(),
            expiredAt: null,
        });
    }
}

File: benchmarks/fixtures/webapp_go/internal/models/session.go
package models

import (
    "fmt"

    "webapp_go/pkg/logger"
)

var sessionLog = logger.GetLogger("models.session")

// Session represents an active user session.
type Session struct {
    ID        string
    UserID    string
    Token     string
    Status    SessionStatus
    IPAddress string
    UserAgent string
    CreatedAt string
    ExpiresAt string
}

// NewSession creates a new session for a user.
func NewSession(userID, token, ip, userAgent string) *Session {
    sessionLog.Info("Creating new session for user: %s", userID)
    return &Session{
        ID:        fmt.Sprintf("sess_%s", userID),
        UserID:    userID,
        Token:     token,
        Status:    SessionActive,
        IPAddress: ip,
        UserAgent: userAgent,
    }
}

// IsValid checks if the session is still active.
func (s *Session) IsValid() bool {
    sessionLog.Debug("Checking session validity: %s", s.ID)
    return s.Status == SessionActive
}

// Expire marks the session as expired.
func (s *Session) Expire() {
    sessionLog.Info("Expiring session: %s", s.ID)
    s.Status = SessionExpired
}

// Revoke marks the session as revoked.
func (s *Session) Revoke() {
    sessionLog.Info("Revoking session: %s", s.ID)
    s.Status = SessionRevoked
}

// Suspend marks the session as suspended.
func (s *Session) Suspend() {
    sessionLog.Info("Suspending session: %s", s.ID)
    s.Status = SessionSuspended
}

// Refresh extends the session expiry.
func (s *Session) Refresh() {
    sessionLog.Info("Refreshing session: %s", s.ID)
    s.Status = SessionActive
}
